ANPR/src/anpr/license_plate_recognition.py

```diff
@@ -0,0 +1,283 @@
+    """
+
+    def __init__(self, selection_ocr: OCR, result_ocr: OCR, formats: list[str] = None, minAR=3,
+                 maxAR=7, verbosity: int = 4, ):
+        # Store the minimum and maximum rectangular aspect ratio
+        self.minAR = minAR
+        self.maxAR = maxAR
+
+        # Variable to store how much debug output the class should give
+        #        Levels of verbosity:
+        #   * -1: nothing
+        #   *  0: prints detected license plates
+        #   *  1: show result image
+        #   *  2: show all detected license plate candidates
+        #   *  3: show result of black/white transformation to find license plate shapes
+        #   *  4: show all steps
+
+        self.verbosity = verbosity
+
+        # Text recognition classes to use for selecting license plates out of multiple candidates and for giving the
+        # final result
+        self.selection_ocr = selection_ocr
+        self.result_ocr = result_ocr
+
+        # Transform custom format to regex ex. 'N-LLL-NNN' -> '[0-9]-[A-Z][A-Z][A-Z]-[0-9][0-9][0-9]'
+        if formats is not None:
+            formats = [f.replace('N', '[0-9]') for f in formats]
+            formats = [f.replace('L', '[A-Z]') for f in formats]
+            self.debug_print(formats, verbosity=0)
+        self.formats = formats
+
+    def debug_imshow(self, title, image, verbosity: int, waitKey=False):
+        """
+        Method to show debug or result images.
+
+
+        @param title: Title of image window
+        @param image: Image to show
+        @param verbosity: level of verbosity required to show the image
+        @param waitKey: boolean indicating whether the program should sleep until the user presses a key.
+        """
+
+        # Check if the verbosity of the image is below the verbosity specified by the user. A low verbosity means the
+        # image is more important
+        if verbosity <= self.verbosity:
+            cv2.imshow(title, image)
+            # Check to see if we should wait for a keypress
+            if waitKey:
+                cv2.waitKey(0)
+
+    def debug_print(self, text, verbosity):
+        """
+        Method print debug or result text
+        @param text: Text to be printed
+        @param verbosity: level of verbosity required to show the image
+        """
+        if verbosity <= self.verbosity:
+            print(text)
+
+    def locate_license_plate_candidates(self, gray, keep=15):
+        # perform a blackhat morphological operation that will allow
+        # us to reveal dark regions (i.e., text) on light backgrounds
+        # (i.e., the license plate itself)
+        rectKern = cv2.getStructuringElement(cv2.MORPH_RECT, (13, 5))
+        blackhat = cv2.morphologyEx(gray, cv2.MORPH_BLACKHAT, rectKern)
+        self.debug_imshow("Blackhat", blackhat, 4)
+
+        # next, find regions in the image that are light
+        squareKern = cv2.getStructuringElement(cv2.MORPH_RECT, (3, 3))
```

Comment:
I think that you copied this from somewhere, because otherwise, brilliant! This is well beyond my cap.